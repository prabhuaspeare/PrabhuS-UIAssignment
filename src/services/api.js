export function fetchTransactions() {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve(require('../transactions.json'));
    }, 1000);
  });
}