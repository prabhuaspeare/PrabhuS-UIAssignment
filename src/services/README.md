# API Module
 
## Overview
The `api.js` module handles fetching transaction data for the application. It simulates an asynchronous API call using a Promise.
 
## Functions
 
### `fetchTransactions()`
- Simulates an API call to retrieve transaction data from a local JSON file.
 
## Usage
1. Import the function:
   ```js
   import { fetchTransactions } from './services/api';