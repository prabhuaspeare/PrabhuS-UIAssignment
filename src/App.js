import React, { useState, useEffect, useMemo } from 'react';
import { fetchTransactions } from './services/api';
import { getCustomerRewards } from './utils/rewardCalculator';
 
function App() {
  const [customers, setCustomers] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);
 
  useEffect(() => {
    async function loadData() {
      try {
        const data = await fetchTransactions();
        setCustomers(data);
      } catch (err) {
        setError(err.message);
      } finally {
        setLoading(false);
      }
    }
 
    loadData();
  }, []);
 
  const customerRewardsData = useMemo(() => {
    return customers.map(customer => ({
      name: customer.name,
      rewards: getCustomerRewards(customer.transactions)
    }));
  }, [customers]);
 
  if (loading) {
    return <div data-testid="loading">Loading...</div>;
  }
 
  if (error) {
    return <div>Error: {error}</div>;
  }
 
  return (
    <div>
      <h1 data-testid="customer-rewards-title">Customer Rewards</h1>
      <table>
        <thead>
          <tr>
            <th>Customer Name</th>
            <th>Date</th>
            <th>Amount</th>
            <th>Points</th>
          </tr>
        </thead>
        <tbody>
          {customerRewardsData.map((customer, customerIndex) =>
            customer.rewards.map((transaction, index) => (
              <CustomerRow
                key={`${customerIndex}-${index}`}
                name={customer.name}
                transaction={transaction}
              />
            ))
          )}
        </tbody>
      </table>
    </div>
  );
}
 
const CustomerRow = React.memo(({ name, transaction }) => (
  <tr>
    <td>{name}</td>
    <td>{transaction.date}</td>
    <td>${transaction.amount}</td>
    <td>{transaction.points}</td>
  </tr>
));
 
export default App;