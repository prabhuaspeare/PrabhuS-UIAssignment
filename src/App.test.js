import { render, screen } from '@testing-library/react';
import App from './App';
import * as api from './services/api';
 
// Mock the entire api module
jest.mock('./services/api');
 
describe('App Component Tests', () => {
    // Test for loading state
    test('renders the loading state initially', () => {
        api.fetchTransactions.mockImplementation(() => new Promise(() => {})); // Mock fetch to keep loading state
        render(<App />);
        const loadingElement = screen.getByTestId('loading');
        expect(loadingElement).toBeInTheDocument();
        expect(loadingElement).toHaveTextContent("Loading...");
    });
 
    // Test for error message
    test('renders error message when fetch fails', async () => {
        api.fetchTransactions.mockImplementationOnce(() => Promise.reject(new Error("Failed to fetch")));
 
        render(<App />);
        const loadingElement = screen.getByTestId('loading');
        expect(loadingElement).toBeInTheDocument();
 
        await screen.findByText(/Error: Failed to fetch/i); // Wait for error message to appear
        expect(screen.queryByTestId('loading')).not.toBeInTheDocument(); // Check loading is not visible anymore
    });
 
    // Test for successful data rendering
    test('renders customer rewards title when fetch succeeds', async () => {
        const mockData = [
            {
                name: "Prasanth",
                transactions: [
                    { date: "2024-01-01", amount: 100, points: 10 },
                ],
            },
        ];
 
        api.fetchTransactions.mockImplementationOnce(() => Promise.resolve(mockData));
 
        render(<App />);
 
        await screen.findByTestId('customer-rewards-title'); // Wait for title to appear
        expect(screen.getByTestId('customer-rewards-title')).toBeInTheDocument();
        expect(screen.getByTestId('customer-rewards-title')).toHaveTextContent("Customer Rewards");
    });
});