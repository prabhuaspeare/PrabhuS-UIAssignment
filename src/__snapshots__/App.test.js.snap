// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component matches snapshot 1`] = `
<DocumentFragment>
  <div>
    Loading...
  </div>
</DocumentFragment>
`;
